File: whatabook/ravenclaw_whatabook.py
""" 
    Title: ravenclaw_whatabook.py
    Author: Hannah Del Real and Tiffany Reyes
    Date: 14 July 2023
    Description: Connecting Python with What-a-Book database on MongoDB
"""

# Import mongoDBClient
from pymongo import MongoClient

# Create a connection string to connect to
client = MongoClient("mongodb+srv://web335_user:s3cret@bellevueuniversity.ozktyyu.mongodb.net/web335DBretryWrites=true&w=majority")

# Assign web335 database to variable db
db = client['web335DB']

# Create space between outputs
print("\n")


# Call the find function to display all books in the collection
def display_books(books):
    for book in books:
        print(f"{book['title']} by {book['author']} \n Details: \n  Genre: {book['genre']} \n  bookId: {book['bookId']} \n")

print("---Displaying all books in What-A-Book's collection--- \n")
display_books(db.books.find({}))

# Display books by genre

def display_books_by_genre(selection):
    print("--Select one of the following genres to see books--- ")
    genres = db.books.distinct("genre")
    for genre in genres:
        print(genre)    
    print(f"\n ---Displaying books by the genre: {selection}--- \n")
    display_books(db.books.find({"genre": selection}))

display_books_by_genre("Fantasy")

# Display books by author

def display_books_by_author(selection):
    print("--Select one of the following authors to see books--- ")
    authors = db.books.distinct("author")
    for author in authors:
        print(author)    
    print(f"\n ---Displaying books by the author: {selection}--- \n")
    display_books(db.books.find({"author": selection}))

display_books_by_author("JK Rowling")

# Display books by bookId

def display_books_by_book_id(book_id):
    print(f"\n ---Displaying books by the bookId: {book_id}--- \n")
    display_books(db.books.find({"bookId": book_id}))

display_books_by_book_id("R001")

# Display books in wishlist by customerId
def display_wishlist_by_customer_id(customer_Id):
    print("Enter your customerId to view your wishlist \n")
    customer = db.customers.find_one({"customerId": customer_Id})
    if customer:
        wishlist = db.wishlists.find_one({"customerId": customer_Id}, {"_id": 0, "wishlistbooks": 1})
        wishlist_items = wishlist["wishlistbooks"]
        customer_name = customer["firstName"]
        print(f"----Displaying books in {customer_name}'s Wishlist: --- \n ")
        for items in wishlist_items:
            book_id = items['bookId']
            details = db.books.find({"bookId": book_id})
            for book in details:
                print(f"* {book['title']} \n    by {book['author']} \n ")
    else:
        print("Invalid, this customerId does not exist in our system. Please try again.")
        
   
display_wishlist_by_customer_id("0004")
        
# Add books to a customer's wishlist
def add_wishlist_book_by_customer_id(customer_id):
    print("Select a book to add your wishlist \n")
    display_books(db.books.find({}))
    customer = db.customers.find_one({"customerId": customer_id})
    if customer:
        db.wishlists.update_one({"customerId": customer_id},  { "$addToSet": {
            "wishlistbooks": {
                "bookId": "R001"
            }
        }})
    else:
        print("Invalid, this customerId does not exist in our system. Please try again.")
        
   
add_wishlist_book_by_customer_id("0004")


# Remove book from a customer's wishlist    
def remove_wishlist_book_by_customer_id(customer_id):
    print("Select a book to add your wishlist \n")
    display_books(db.books.find({}))
    customer = db.customers.find_one({"customerId": customer_id})
    if customer:
        db.wishlists.update_one({"customerId": customer_id},  { "$pull": {
            "wishlistbooks": {
                "bookId": { "$eq": "R001" }
            }
        }})
    else:
        print("Invalid, this customerId does not exist in our system. Please try again.")

   
remove_wishlist_book_by_customer_id("0004")


                

File: whatabook/ravenclaw-whatabook.js
/*
===========================================================================================
; Title: ravenclaw-whatabook.js
; Author: Tiffany Reyes and Hannah Del Real
; Date: 11 July 2023
; Description: WhatABook queries
;==========================================================================================
*/

// Delete the books, customers, and wishlists collections.
db.books.drop();
db.customers.drop();
db.wishlists.drop();

// Create the books, customers, and wishlists collections.
db.createCollection("books", {
	validator: { $jsonSchema: {
		bsonType: "object",
        required: ["title", "author", "genre", "bookId"],
		properties: {
			title: {
				bsonType: "string"
			},
			author: {
				bsonType: "string"
			},
            genre: {
				bsonType: "string"
			},
			bookId: {
				bsonType: "string"
			}
		}
	}}
});

db.createCollection("customers", {
	validator: { $jsonSchema: {
		bsonType: "object",
        required: ["customerId", "firstName", "lastName"],
		properties: {
			customerId: {
				bsonType: "string"
			},
			firstName: {
				bsonType: "string"
			},
            lastName: {
				bsonType: "string"
			}
		}
	}}
});

db.createCollection("wishlists", {
	validator: { $jsonSchema: {
		bsonType: "object",
        required: ["wishlistId", "customerId", "wishlistbooks"],
		properties: {
            wishlistId: {
				bsonType: "string"
			},
            customerId: {
				bsonType: "string"
			},
			wishlistbooks: {
				bsonType: "array"
			}
		}
	}}
});

// Books
let book1 = {
    "title": "Pride and Prejudice",
    "author": "Jane Austin",
    "genre": "Romance",
    "bookId": "R001"
};

let book2 = {
    "title": "Harry Potter and the Prisoner of Azakaban",
    "author": "JK Rowling",
    "genre": "Fantasy",
    "bookId": "F003"
};

let book3 = {
    "title": "Harry Potter and the Goblet of Fire",
    "author": "JK Rowling",
    "genre": "Fantasy",
    "bookId": "F004"
};

let book4 = {
    "title": "The Last Song",
    "author": "Nicholas Sparks",
    "genre": "Romance",
    "bookId": "R002"
};
let book5 = {
    "title": "The Postman Always Rings Twice",
    "author": "James M. Cain",
    "genre": "Mystery",
    "bookId": "M008"
};


// Insert book documents
db.books.insertOne(book1);
db.books.insertOne(book2);
db.books.insertOne(book3);
db.books.insertOne(book4);
db.books.insertOne(book5);

// Customers
let customer1 = {
    "customerId": "0003",
    "firstName": "Mina",
    "lastName": "Myoi"
};

let customer2 = {
    "customerId": "0004",
    "firstName": "Jihyo",
    "lastName": "Park"
};

// Insert customer documents
db.customers.insertOne(customer1);
db.customers.insertOne(customer2);

// Wishlists
let wishlist1 = {
    "wishlistId": "10001",
    "customerId": "0004",
    "wishlistbooks": [
        {
            "bookId": "F003"
        },
        {
            "bookId": "M008"
        }
    ]
};

let wishlist2 = {
    "wishlistId": "10000",
    "customerId": "0003",
    "wishlistbooks": [
        {
            "bookId": "R001"
        },
        {
            "bookId": "R002"
        }
    ]
};


// Insert wishlists documents
db.wishlists.insertOne(wishlist1);
db.wishlists.insertOne(wishlist2);

// Query to display list of books
db.books.find();

// Query to display list of books by genre
db.books.find( { "genre": "Fantasy" } );

// Query to display a list of books by author
db.books.find( { "author": "Jane Austin" } );

// Query to display a book by bookId
db.books.find( { "bookId": "R001" } );

// Query to display a wishlist by customerId
db.wishlists.aggregate([{$lookup: {from: "books", localField: "wishlistbooks.bookId", foreignField: "bookId", as: "wishlistbooks"}}, {$match: {"customerId": "0004"}}])

// Query to add a book to customer's wishlist
db.wishlist.updateOne(
    { "customerId": "0004" },
    { "$addToSet": {
        wishlistbooks: {
            "bookId": "R001"
        }
    }}
);

// Query to remove a book from customer's wishlist
db.wishlist.updateOne(
    { "customerId": "0004" },
    { "$pull": {
        wishlistbooks: { "bookId": { $eq: "R001" } }
    }
}
);
